src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+                delay_start_adc = int(
+                    int(
+                        np.ceil(
+                            self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9 / 8
+                        )
+                        / 2
+                    )
+                    * 2
+                )
 
-    def stop(self):
-        self._send(b"stop")
+                # For the DAC, currently the sampling rate is 3x higher than the ADC
+                # The number of clock cycles is 16 samples per clock cycle
+                # Hence, we multiply the adc delay clock cycles by 1.5x to align the DAC/ADC pair
+                delay_start_dac = int(delay_start_adc * 1.5)
 
-    def status(self):
-        return self._send(b"status", True)
+                self.device.dac[dac].delay = (
+                    delay_start_dac + self.channel_delay_offset_dac
+                )
+                self.device.adc[adc].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
+                # ADC0 complete marks the end of acquisition, so we also need to move ADC0
+                self.device.adc[0].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
 
-    def _send(self, payload, retval=False):
-        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-            s.connect((self.address, self.port))
-            s.sendall(payload)
+                if (
+                    options.acquisition_type is AcquisitionType.DISCRIMINATION
+                    or AcquisitionType.INTEGRATION
+                ):
+                    self.device.program_qunit(
+                        readout_frequency=pulse.frequency,
+                        readout_time=pulse.duration * 1e-9,
+                        qunit=pulse.qubit,
+                    )
+
+            end = start + len(wfm)
+            waveform_array[dac][start:end] += self.device.dac_max_amplitude * wfm
+            dac_end_addr[dac] = max(end >> 4, dac_end_addr[dac])
+
+        payload = [
+            (dac, wfm, dac_end_addr[dac])
+            for dac, wfm in waveform_array.items()
+            if dac_end_addr[dac] != 0
+        ]
+        self.device.upload_waveform(payload)
+
+    def play_sequences(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+    ):
+        pass
 
-            if retval:
-                return s.recv(1024).decode("utf-8")
+    def connect(self):
+        """Currently we only connect to the board when we have to send a
+        command."""
+        # Request the version from the board
+        ver = self.device.get_server_version()
+        log.info(f"Connected to {self.name}, version: {ver}")
 
-    def play_sequence(self):
-        self.fire()
+    def start(self):
+        pass
 
-    def connect(self):
+    def stop(self):
```

Comment:
I noticed that following these comments `start()` and `stop()` are now dropped. Note that this will not work because they are still `@abstractmethod` of the `Instrument` in this branch, so you'll probably get an error that some abstract methods are missing when you try to initialize your instruments.

@AleCandido shall we merge #739 before this to fix this issue? I will try to merge it today. It already has the approvals, I just want to check the QPU tests.

Reply:
> @AleCandido shall we merge #739 before this PR, to fix this issue? I will try to merge it today. It already has the approvals, I just want to check the QPU tests.

I'd rather try to merge #739 anyhow, because we won't have a fridge for a while.
The bulk of the changes in #739 is not heavily dependent on the devices, so if the tests we have are sensible enough[*] then I'd just merge it.

[*]: of course we don't know, but we don't even know for the QPU ones...

Reply:
I expect this to be merged on Monday or right after, anyhow. So, *today* would be enough :)

(expectations based on time zones)

Reply:
#739 is merged, so now this should be fine after merging main to it.

Reply:
Thanks @stavros11